core/src/main/java/org/elasticsearch/common/settings/AbstractScopedSettings.java

```diff
@@ -519,7 +519,14 @@ public Settings archiveUnknownOrInvalidSettings(
             try {
                 Setting<?> setting = get(entry.getKey());
                 if (setting != null) {
-                    setting.get(settings);
+                    try {
+                        setting.get(settings);
+                    } catch (IllegalArgumentException ex) {
+                        if (setting.isMandatory()) {
+                            throw new IllegalStateException("can't archive mandatory setting [" + setting.getKey() + "]", ex);
```

Comment:
People don't know what archiving setting means. They just upgrade and their end up in an illegal state. I think it will be cleared not to mention the "can't be archived" part. But as said - just a nit, not a big deal.